### PR TITLE
Adds the ability to group within the multiselect

### PR DESCRIPTION
**Pull request by gabrielmiller**

### CODE REVIEW

File: src/app/multiselect/multiselect.factory.js [hunk 1]
```diff
@@ -13,7 +13,14 @@
      */
     function MultiselectFactory($parse) {
 
-        var _optionsRegularExpression = /^\s*(?:(\S+)\s+as\s+)?(\S+)\s+for\s+(\S+)\s+in\s+(\S+)\s*$/;
+        var _optionsRegularExpression = /^\s*(?:(\S+)\s+as\s+)?(\S+)(?:\s+group\s+by\s+([\s\S]+?))?\s+for\s+(\S+)\s+in\s+(\S+)\s*$/;
```
Comment (namoscato) [inline src/app/multiselect/multiselect.factory.js]: Is there a reason why the character class (`[\s\S]`) for the group includes whitespace?

File: src/app/multiselect/multiselect.filter.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+(function() {
+    'use strict';
+
+    angular
+        .module('amo.multiselect')
+        .filter('amoMultiselect', AmoMultiselectFilter);
```
Comment (namoscato) [inline src/app/multiselect/multiselect.filter.js]: Could we just name this `amoMultiselectGroup` and omit the `method` argument?

File: src/app/multiselect/multiselect.directive.js [hunk 3]
```diff
@@ -100,14 +106,26 @@
...
+                    } else {
+                        optionObj.group = 'ungrouped';
+                    }
+
+                    _groupsHash[optionObj.group] = true;
+                    self.options.push(optionObj);
                 });
 
+                // Set the groups array
+                self.groups = Object.keys(_groupsHash);
```
Comment (namoscato) [inline src/app/multiselect/multiselect.directive.js]: Do you think the order matters here?

File: src/app/multiselect/multiselect-dropdown.html [hunk 4]
```diff
@@ -27,13 +27,15 @@
                     ng-click="multiselectDropdown.deselectAll()">
                 </a>
             </li>
-            <li class="divider"></li>
-            <li ng-repeat="option in multiselectDropdown.optionsFiltered = (multiselectDropdown.options | filter : multiselectDropdown.search)">
+            <li ng-bind="group" ng-if="multiselectDropdown.isGrouped && !multiselectDropdown.isGroupEmpty(group)" ng-repeat-start="group in multiselectDropdown.groups"></li>
+            <li class="divider" ng-if="!multiselectDropdown.isGrouped"></li>
```
Comment (namoscato) [inline src/app/multiselect/multiselect-dropdown.html]: I would probably just always include the divider before the outer `ngRepeat`. Also, we can throw a `dropdown-header` class on the group header list item.

File: src/app/multiselect/multiselect.directive.js [hunk 5]
```diff
@@ -186,6 +204,17 @@
...
 
+
+            /**
+             * @ngdoc method
+             * @name amoMultiselect#isGroupEmpty
+             * @description Determines whether or not items are selected
+             * @returns {Boolean}
+             */
+            function isGroupEmpty(group) {
+                return $filter('amoMultiselect')($filter('filter')(self.options, self.search), 'grouping', group).length === 0;
```
Comment (namoscato) [inline src/app/multiselect/multiselect.directive.js]: Minor, but you can inject `filterFilter` and `amoMultiselectFilter` – which reference the resulting filter functions directly.

**Comment by namoscato** [comment]
Nice job with this man!

File: src/app/multiselect/multiselect.factory.js (hunk 1 above)
Comment (gabrielmiller) [inline src/app/multiselect/multiselect.factory.js]: I just duplicated this from the regex in ng-options. This is a good point though. I don't think the space is necessary

File: src/app/multiselect/multiselect.directive.js (hunk 3 above)
Comment (gabrielmiller) [inline src/app/multiselect/multiselect.directive.js]: Good point. I am going to assume we should retain the order of the groups as they come up in the options.

File: src/app/multiselect/multiselect.filter.js (hunk 2 above)
Comment (gabrielmiller) [inline src/app/multiselect/multiselect.filter.js]: Yep.

File: src/app/multiselect/multiselect.directive.js (hunk 5 above)
Comment (gabrielmiller) [inline src/app/multiselect/multiselect.directive.js]: Sure, but is there any benefit to doing it that way versus just using ```$filter``` to fetch both filters?

Comment (namoscato) [inline src/app/multiselect/multiselect.directive.js]: Nope – just arguably easier to test.

**Pull request merged**